packages/blockchain/src/consensus/clique.ts

```diff
@@ -0,0 +1,557 @@
+  }
+
+  /**
+   * Checks if signer was recently signed.
+   * Returns true if signed too recently: more than once per {@link Blockchain.cliqueSignerLimit} consecutive blocks.
+   * @param header BlockHeader
+   * @hidden
+   */
+  private cliqueCheckRecentlySigned(header: BlockHeader): boolean {
+    if (header.isGenesis() || header.number === BigInt(1)) {
+      // skip genesis, first block
+      return false
+    }
+    const limit = this.cliqueSignerLimit()
+    // construct recent block signers list with this block
+    let signers = this._cliqueLatestBlockSigners
+    signers = signers.slice(signers.length < limit ? 0 : 1)
+    signers.push([header.number, header.cliqueSigner()])
+    const seen = signers.filter((s) => s[1].equals(header.cliqueSigner())).length
+    return seen > 1
+  }
+
+  /**
+   * Remove clique snapshots with blockNumber higher than input.
+   * @param blockNumber - the block number from which we start deleting
+   * @hidden
+   */
+  private async _cliqueDeleteSnapshots(blockNumber: bigint) {
+    // remove blockNumber from clique snapshots
+    // (latest signer states, latest votes, latest block signers)
+    this._cliqueLatestSignerStates = this._cliqueLatestSignerStates.filter(
+      (s) => s[0] <= blockNumber
+    )
+    await this.cliqueUpdateSignerStates()
+
+    this._cliqueLatestVotes = this._cliqueLatestVotes.filter((v) => v[0] <= blockNumber)
+    await this.cliqueUpdateVotes()
+
+    this._cliqueLatestBlockSigners = this._cliqueLatestBlockSigners.filter(
+      (s) => s[0] <= blockNumber
+    )
+    await this.cliqueUpdateLatestBlockSigners()
+  }
+
+  /**
+   * Update snapshot of latest clique block signers.
+   * Used for checking for 'recently signed' error.
+   * Length trimmed to {@link Blockchain.cliqueSignerLimit}.
+   * @param header BlockHeader
+   * @hidden
+   */
+  private async cliqueUpdateLatestBlockSigners(header?: BlockHeader) {
+    if (header) {
+      if (header.isGenesis()) {
+        return
+      }
+      // add this block's signer
+      const signer: CliqueBlockSigner = [header.number, header.cliqueSigner()]
+      this._cliqueLatestBlockSigners.push(signer)
+
+      // trim length to `this.cliqueSignerLimit()`
+      const length = this._cliqueLatestBlockSigners.length
+      const limit = this.cliqueSignerLimit()
+      if (length > limit) {
+        this._cliqueLatestBlockSigners = this._cliqueLatestBlockSigners.slice(
+          length - limit,
+          length
+        )
+      }
+    }
+
+    // save to db
+    const formatted = this._cliqueLatestBlockSigners.map((b) => [
+      bigIntToBuffer(b[0]),
+      b[1].toBuffer(),
+    ])
+    await this.blockchain.db.put('CliqueBlockSignersSnapshot', rlp.encode(formatted), DB_OPTS)
+  }
+
+  /**
+   * Fetches clique signers.
+   * @hidden
+   */
+  private async getCliqueLatestSignerStates(): Promise<CliqueLatestSignerStates> {
+    try {
+      const signerStates = await this.blockchain.db.get('CliqueSigners', DB_OPTS)
```

Comment:
you should use the constant DB keys here (`CLIQUE_SIGNERS_KEY`, `CLIQUE_VOTES_KEY`, `CLIQUE_BLOCK_SIGNERS_KEY`) rather than string values

Reply:
Holger was advocating for not using dbManager and instead just simplifying the db interactions as much as possible in the clique class, so that's why it's not currently using the DBTarget pattern. But also, agreed, it's better to define the strings once as constants. Updated